WalletWasabi/Stores/BlockFilterSqliteStorage.cs

```diff
@@ -0,0 +1,249 @@
+using Microsoft.Data.Sqlite;
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using WalletWasabi.Backend.Models;
+
+namespace WalletWasabi.Stores;
+
+/// <summary>
+/// Sqlite-based storage for block filters.
+/// </summary>
+/// <remarks>The implementation is not thread-safe because <see cref="SqliteConnection"/> is not thread-safe.</remarks>
+/// <seealso href="https://learn.microsoft.com/en-gb/dotnet/standard/data/sqlite/async">
+/// Async is not supported at the moment, so no async methods were used in this implementation.
+/// </seealso>
+public class BlockFilterSqliteStorage : IDisposable
+{
+	/// <seealso href="https://learn.microsoft.com/en-us/dotnet/standard/data/sqlite/in-memory-databases"/>
+	public const string InMemoryDatabase = ":memory:";
+
+	private bool _disposedValue;
+
+	private BlockFilterSqliteStorage(SqliteConnection connection)
+	{
+		Connection = connection;
+	}
+
+	/// <remarks>Connection cannot be accessed from multiple threads at the same time.</remarks>
+	private SqliteConnection Connection { get; }
+
+	/// <summary>
+	/// Opens a new sqlite connection to the given database file.
+	/// </summary>
+	/// <param name="dataSource">Path to the sqlite database file, or special <c>:memory:</c> string.</param>
+	/// <param name="startingFilter">Starting filter to put into the filter table if the table needs to be created.</param>
+	/// <exception cref="InvalidOperationException">If there is an unrecoverable error.</exception>
+	/// <seealso href="https://dev.to/lefebvre/speed-up-sqlite-with-write-ahead-logging-wal-do">Write-ahead logging explained.</seealso>
+	public static BlockFilterSqliteStorage FromFile(string dataSource, FilterModel? startingFilter = null)
+	{
+		SqliteConnectionStringBuilder builder = new();
+		builder.DataSource = dataSource;
+
+		SqliteConnection connection = new(builder.ConnectionString);
+		connection.Open();
+
+		using (SqliteCommand createCommand = connection.CreateCommand())
+		{
+			createCommand.CommandText = """
+				CREATE TABLE IF NOT EXISTS filter (
+					block_height INTEGER NOT NULL PRIMARY KEY,
+					block_hash BLOB NOT NULL,
+					filter_data BLOB NOT NULL,
+					previous_block_hash BLOB NOT NULL,
+					epoch_block_time INTEGER NOT NULL
+				)
+				""";
+			createCommand.ExecuteNonQuery();
+		}
+
+		// Enable write-ahead logging.
+		using (SqliteCommand walCommand = connection.CreateCommand())
+		{
+			walCommand.CommandText = """
+				PRAGMA journal_mode = 'wal';
+				PRAGMA synchronous  = 'NORMAL';
+				""";
+			walCommand.ExecuteNonQuery();
+		}
+
+		BlockFilterSqliteStorage storage = new(connection);
+
+		if (startingFilter is not null)
+		{
+			using SqliteCommand isEmptyCommand = connection.CreateCommand();
+			isEmptyCommand.CommandText = "SELECT count(*) FROM filter";
```

Comment:
Suggestion for consistency, in other places you are using uppercase SQL function naming (like for `MAX()`).
```suggestion
			isEmptyCommand.CommandText = "SELECT COUNT(*) FROM filter";
```